rfc-171-remove-legacy-browser-js-support.md

```diff
@@ -0,0 +1,106 @@
+status: proposed
+implementation: proposed
+status_last_reviewed:
+---
+
+# Remove JavaScript support for legacy browsers
+
+## Summary
+
+This RFC proposes removing JavaScript (JS) support for old (legacy) browsers such as Internet Explorer across the GOV.UK website. This is because:
+
+- the browsers in question are no longer supported by their vendors
+- we cannot offer these browsers an entirely equal experience because `govuk-frontend` has already dropped JS support for legacy browsers
+- the percentage of users likely to be affected is well below 1%, and will not increase
+- we build progressively and the site will remain legible, navigable and usable for all users regardless of this change
+
+The benefits of this are:
+
+- reduced burden on developers to write extra code to support legacy browsers
+- minor performance improvements for other users as we stop branching code and including polyfills
+- we can embrace modern JS techniques, which could improve and simplify our code
+
+The drawbacks are:
+
+- users of legacy browsers will no longer be included in our analytics tracking
+
+## Problem
+
+This RFC is taking a broad approach to legacy browsers, but the main point of contention is likely to be Internet Explorer 11, the legacy browser with the most significant number of ongoing visitors.
+
+At the time of writing the service manual (which was last updated in 2022) says "You will still need to support Internet Explorer 11 if your analytics data shows at least 2% of your users arriving at the service's start page are using it." It [recommends testing](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) only in Edge of all the Microsoft browsers.
+
+The Design System team have updated `govuk-frontend` to use [ES modules in version 5](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#javascript), released in 2023. [ES modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) are not recognised by legacy browsers, notably Internet Explorer 11 and below. These browsers are no longer supported by Microsoft. We need to keep up to date with `govuk-frontend`, which means deciding how we approach JS support for IE11.
+
+Legacy browsers will error if they attempt to parse or run any JS syntax that they do not recognise.
+
+## Proposal
+
+The proposal is to modify GOV.UK so that legacy browsers do not run any JS on GOV.UK. This should still provide a usable experience as the site is built using the principle of progressive enhancement. This will ultimately reduce page weight for those browsers and allow us to embrace modern JS techniques, leading to further code improvements in future.
+
+This will be achieved by switching our JS script tags to include `type="module"`. This is used in modern browsers to differentiate JS containing ES modules. Legacy browsers do not recognise this attribute and may download the JS but will not attempt to parse or run it.
```

Comment:
Looks good to me, a couple of other things for us to consider when it comes to implementation:

- We will likely want to follow the same approach used in v5 of govuk-frontend in our own modules as well - https://github.com/alphagov/govuk-frontend/pull/3801
- By default, when `type=module` attribute is present, the script will automatically be deferred, for example, the scripts below run without waiting for any other events to fire, in this case the scripts just log the application name to the console.

```HTML
<script src="/assets/static/application.js"></script>
<script src="/assets/frontend/application.js"></script>
```

Console output
```JS
"static"
"frontend"
```

**Using type=module**

```HTML
<script src="/assets/static/application.js" type="module"></script>
<script src="/assets/frontend/application.js"></script>
```

Console output
```JS
"frontend"
"static"
```